src/js.c

```diff
@@ -2197,38 +2197,68 @@ _checkConfig(jsConsumerConfig *s, jsConsumerConfig *u)
 
 static natsStatus
 _processConsInfo(const char **dlvSubject, jsConsumerInfo *info, jsConsumerConfig *userCfg,
-                 bool isPullMode, const char *subj, const char *queue)
+                 bool isPullMode, const char **subjects, int numSubjects, const char *queue)
 {
-    bool                dlvSubjEmpty = false;
-    jsConsumerConfig    *ccfg        = info->Config;
-    const char          *dg          = NULL;
-    natsStatus          s            = NATS_OK;
-    bool                matches      = false;
-    int                 i;
+    bool dlvSubjEmpty = false;
+    jsConsumerConfig *ccfg = info->Config;
+    const char *dg = NULL;
+    natsStatus s = NATS_OK;
+    const char *stackFilterSubject[] = {ccfg->FilterSubject};
+    const char **filterSubjects = stackFilterSubject;
+    int filterSubjectsLen = 1;
+    int incoming, existing;
 
     *dlvSubject = NULL;
+    // Always represent the consumer's filter subjects as a list, to match
+    // uniformly against the incoming subject list. Consider lists of 1 empty
+    // subject empty lists.
+    if (ccfg->FilterSubjectsLen > 0)
+    {
+        filterSubjects = ccfg->FilterSubjects;
+        filterSubjectsLen = ccfg->FilterSubjectsLen;
+    }
+    if ((filterSubjectsLen == 1) && nats_IsStringEmpty(filterSubjects[0]))
+    {
+        filterSubjects = NULL;
+        filterSubjectsLen = 0;
+    }
+    if ((numSubjects == 1) && nats_IsStringEmpty(subjects[0]))
+    {
+        subjects = NULL;
+        numSubjects = 0;
+    }
 
-    // Make sure this new subject matches or is a subset.
-    if (!nats_IsStringEmpty(subj))
+    // Match the subjects against the consumer's filter subjects.
+    if (numSubjects > 0 && filterSubjectsLen > 0)
     {
-        if (nats_IsStringEmpty(ccfg->FilterSubject) && (ccfg->FilterSubjectsLen == 0))
```

Comment:
This was done incorrectly before this change, now restore the prior logic of matching exactly only if both subjects and filter subject(s) are set.